docs/build_and_push_docker_image_workflow_documentation.md

```diff
@@ -0,0 +1,219 @@
+
+### Permissions
+
+```yaml
+permissions:
+  contents: read    # Allows read access to the repository code
+  packages: write   # Allows pushing images to GitHub Container Registry
+```
+
+- **Contents**: Read access to fetch the repository code.
+- **Packages**: Write access to push Docker images to GHCR.
+
+## Job Breakdown
+
+### Job: `build-and-push`
+
+#### Runs On
+
+- **Environment**: `ubuntu-latest`
+
+#### Steps
+
+1. **Checkout Code**
+
+   ```yaml
+   - name: Checkout code
+     uses: actions/checkout@v4
+   ```
+
+   - **Description**: Checks out the repository code so that the workflow can access it.
+
+2. **Set Up Docker Buildx**
+
+   ```yaml
+   - name: Set up Docker Buildx
+     uses: docker/setup-buildx-action@v3
+   ```
+
+   - **Description**: Sets up Docker Buildx for building multi-platform images.
+
+3. **Log in to GitHub Container Registry**
+
+   ```yaml
+   - name: Log in to GitHub Container Registry
+     uses: docker/login-action@v3
+     with:
+       registry: ghcr.io
+       username: ${{ github.actor }}
+       password: ${{ secrets.GITHUB_TOKEN }}
+   ```
+
+   - **Description**: Authenticates with GHCR using your GitHub credentials.
+
+4. **Determine the Release Tag**
+
+   ```yaml
+   - name: Determine the release tag
+     id: get_tag
+     run: |
+       if [ "${{ github.event_name }}" == "release" ]; then
+         echo "tag=${GITHUB_REF#refs/tags/}" >> $GITHUB_OUTPUT
+       else
+         echo "tag=${{ github.event.inputs.release_tag }}" >> $GITHUB_OUTPUT
+       fi
+   ```
+
+   - **Description**: Determines the Docker image tag:
+     - If triggered by a release, uses the release tag.
+     - If manually triggered, uses the provided `release_tag` input.
+
+5. **Build and Push Docker Image**
+
+   ```yaml
+   - name: Build and push Docker image
+     uses: docker/build-push-action@v6
+     with:
+       context: .
+       push: true
+       tags: ghcr.io/${{ github.repository_owner }}/${{ github.event.repository.name }}:${{ steps.get_tag.outputs.tag }}
+       platforms: linux/arm64,linux/amd64
+   ```
+
+   - **Description**: Builds and pushes the Docker image to GHCR for the specified platforms.
+
+## Platforms
+
+The workflow builds Docker images for the following platforms:
+
+- **linux/arm64**
+- **linux/amd64**
+
+**Note**: Initially, the aim was to build for additional platforms such as `darwin/amd64` and `windows/amd64`, but due to local limitations, the workflow currently builds for the platforms listed above.
+
+## Requirements and Prerequisites
+
+To effectively use this workflow, ensure the following:
+
+- **Dockerfile**: A valid `Dockerfile` must be present at the root of your repository.
+- **GitHub Permissions**: The `GITHUB_TOKEN` provided by GitHub Actions must have the necessary permissions (default settings typically suffice).
+- **Understanding of Docker and GHCR**: Basic knowledge of Docker image building and pushing to GHCR will be beneficial.
```

Comment:
This was the core of what I was wanting to understand.

If I get the gist of it, instead of us pushing to docker hub, we are pushing to github's container registry which allows us to just use GH_TOKEN without having to setup any count, register secrets, and/or do any manual per-release action.

Is that right?